runners/flink/src/main/java/org/apache/beam/runners/flink/metrics/FlinkMetricContainer.java

```diff
@@ -94,121 +93,128 @@ public MetricsContainer getMetricsContainer(String stepName) {
         : null;
   }
 
+  public MetricsContainer getUnboundMetricsContainer() {
+    return metricsAccumulator != null
+        ? metricsAccumulator.getLocalValue().getUnboundContainer()
+        : null;
+  }
+
   /**
    * Update this container with metrics from the passed {@link MonitoringInfo}s, and send updates
    * along to Flink's internal metrics framework.
    */
-  public void updateMetrics(String stepName, List<MonitoringInfo> monitoringInfos) {
-    MetricsContainer metricsContainer = getMetricsContainer(stepName);
+  public void updateMetrics(List<MonitoringInfo> monitoringInfos) {
```

Comment:
Question about the overall design here. Much of this class doesn't seem flink specific. If its just aggregating the values in memory, perhaps it should just aggregate the MonitoringInfos. Then have some method to extract them in a flink specific format (possibly in another class)

Reply:
Interesting point. To paraphrase: ingesting MonitoringInfos into a MetricContainerStepMap (which other runners also use) will be a common path, and code for that should get factored out.

I experimented with keeping fn-api metrics natively in MonitoringInfos in MCSM, alongside traditional Java-SDK-formatted ones, but ended up needing to do a lot of other cleanup in there (#7890) before I could see straight 😄.